controlplane/kubeadm/internal/filters.go

```diff
@@ -29,101 +31,136 @@ import (
+	rolloutReasons := []string{}
+
+	// Machines whose certificates are about to expire.
+	if collections.ShouldRolloutBefore(reconciliationTime, rolloutBefore)(machine) {
+		rolloutReasons = append(rolloutReasons, "certificates will expire soon, rolloutBefore expired")
+	}
+
+	// Machines that are scheduled for rollout (KCP.Spec.RolloutAfter set,
+	// the RolloutAfter deadline is expired, and the machine was created before the deadline).
+	if collections.ShouldRolloutAfter(reconciliationTime, rolloutAfter)(machine) {
+		rolloutReasons = append(rolloutReasons, "rolloutAfter expired")
+	}
+
+	// Machines that do not match with KCP config.
+	if mismatchReason, matches := matchesMachineSpec(infraConfigs, machineConfigs, kcp, machine); !matches {
+		rolloutReasons = append(rolloutReasons, mismatchReason)
+	}
+
+	if len(rolloutReasons) > 0 {
+		return fmt.Sprintf("Machine %s needs rollout: %s", machine.Name, strings.Join(rolloutReasons, ",")), true
+	}
+
+	return "", false
 }
 
-// MatchesTemplateClonedFrom returns a filter to find all machines that have a corresponding infrastructure machine that
+// matchesTemplateClonedFrom returns a filter to find all machines that have a corresponding infrastructure machine that
 // matches a given KCP infra template.
 // Note: Differences to the labels and annotations on the infrastructure machine are not considered for matching
 // criteria, because changes to labels and annotations are propagated in-place to the infrastructure machines.
 // TODO: This function will be renamed in a follow-up PR to something better. (ex: MatchesInfraMachine).
-func MatchesTemplateClonedFrom(infraConfigs map[string]*unstructured.Unstructured, kcp *controlplanev1.KubeadmControlPlane) collections.Func {
-	return func(machine *clusterv1.Machine) bool {
-		if machine == nil {
-			return false
-		}
-		infraObj, found := infraConfigs[machine.Name]
-		if !found {
-			// Return true here because failing to get infrastructure machine should not be considered as unmatching.
-			return true
-		}
-
-		clonedFromName, ok1 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
-		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
-		if !ok1 || !ok2 {
-			// All kcp cloned infra machines should have this annotation.
-			// Missing the annotation may be due to older version machines or adopted machines.
-			// Should not be considered as mismatch.
-			return true
-		}
+func matchesTemplateClonedFrom(infraConfigs map[string]*unstructured.Unstructured, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool) {
+	if machine == nil {
+		return "Machine cannot be compared with KCP.spec.machineTemplate.infrastructureRef: Machine is nil", false
```

Comment:
just curious, is this an actual expected scenario or a code guard jic?

Reply:
Should never happen